### PR TITLE
    PowerPoint: Report link destination

### PR DESCRIPTION
**Pull request by CyrilleB79**
### Link to issue number:
None
### Summary of the issue:
When in PowerPoint, NVDA+K does not work to report a link.

### Description of user facing changes
* In PowerPoint slides (edition mode), pressing NVDA+K will now report the link destination. This works in text areas as well as on shapes / charts / images that are links.
* Also 2 small fixes:
  * Restored the Copy and Close buttons of the browseable message of NVDA+K that were removed by mistake in #17292
  * In Windows UI, e.g. settings, pressing NVDA+K on a link does not report anymore that there is no link; it rather reports that the destination cannot be reported.

### Description of development approach
* Use PowerPoint object model
* Changed the global strategy to retrieve a link: first try from text infos; if not supported, try from the focused object.

### Testing strategy:
* Tested PowerPoint link reporting in text area, on shapes and images
* Re-tested link reporting in browsers (Chrome, FF and `ui.browseableMessage`), in Word / Outlook with/without UIA (text and images), in Excel.

### Known issues with pull request:
* NVDA+K doesn't still work in the slideshow; by the way, the presence of links is not reported either in slideshow.
* There are other situations where link is not reported by NVDA+K on the web or elsewhere (WordPad, Adobe Reader) but this PR does not degrade any experience.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

@coderabbitai summary

### CODE REVIEW

**Comment by CyrilleB79** [comment]
For reviewers (@SaschaCowley or @seanbudd ):

OK finally, I'll mark this as ready.

My initial intention was to rework the code to avoid reporting "Not a link" on links in applications where retrieving the destination of a link is not implemented (e.g. Adobe Reader). But I am unable to create a function to detect if the caret is on a link. Since it was a bonus in this PR,I won't keep it blocked on this point.

Also, I'd like to move the class definition for `_LinkData` class elsewhere than in `textInfos.__init__.py` since it does not apply anymore only to text infos. But I do not know where to move it. Any suggestion?

**Comment by seanbudd** [comment]
@CyrilleB79 my suggestion from `_LinkData` is `utils.urlUtils`

**Comment by CyrilleB79** [comment]
> @CyrilleB79 my suggestion from `_LinkData` is `utils.urlUtils`

OK thanks. Done.

File: source/globalCommands.py [hunk 1]
```diff
@@ -4280,16 +4280,19 @@ def script_reportLinkDestination(
 		positioned on a link, or an element with an included link such as a graphic.
 		:param forceBrowseable: skips the press once check, and displays the browseableMessage version.
 		"""
+		focus = api.getFocusObject()
 		try:
 			ti: textInfos.TextInfo = api.getCaretPosition()
+			link = ti._getLinkDataAtCaretPosition()
```
Comment (seanbudd) [inline source/globalCommands.py:4286]: is this expected to raise a RuntimeError? If not, it should be moved to an `else` block below

File: source/globalCommands.py [hunk 2]
```diff
@@ -4304,18 +4307,19 @@ def script_reportLinkDestination(
 					link.destination,
 					# Translators: Informs the user that the window contains the destination of the
 					# link with given title
-					title=_("Destination of: {name}").format(
-						name=text,
-						closeButton=True,
-						copyButton=True,
```
Comment (seanbudd) [inline source/globalCommands.py:4310]: this looks like a strange logic error. is there a known bug associated with this?

File: source/globalCommands.py [hunk 3]
```diff
@@ -4280,16 +4280,19 @@ def script_reportLinkDestination(
...
+			try:
+				link = focus.linkData
+			except NotImplementedError:
+				link = None
 		presses = scriptHandler.getLastScriptRepeatCount()
 		if link:
 			if link.destination is None:
-				# Translators: Informs the user that the link has no destination
+				# Translators: Reported when using the command to report the destination of a link.
 				ui.message(_("Link has no apparent destination"))
```
Comment (seanbudd) [inline source/globalCommands.py:4296]: Apparent seems strange here. Would this change be misleading?

```suggestion
				ui.message(_("Link has no destination"))
```

Comment (CyrilleB79) [reply source/globalCommands.py:4296]: This text was introduced in #14723 by @lukaszgo1 on a suggestion of @XLTechie (https://github.com/nvaccess/nvda/pull/14723#issuecomment-1480802046).

@CyrilleB79 what is your opinion? Maybe a rewording would be needed.

But from what I understand, the link may have a destination; it's just that we are not able to determine it.

File: source/globalCommands.py (hunk 2 above)
Comment (CyrilleB79) [reply source/globalCommands.py:4310]: Are you referring to the following (in initial description of this PR)?
> ◦ Restored the Copy and Close buttons of the browseable message of NVDA+K that were removed by mistake in Report destination of link in Word / Outlook legacy #17292

If yes, the issue has not been logged anywhere else in GitHub; it's a regression of my own previous PR in the same dev cycle.

File: source/globalCommands.py (hunk 1 above)
Comment (CyrilleB79) [reply source/globalCommands.py]: Fixed in b019bd162ce7778948c9ec89ff62a6e9cea976b7

**Review by seanbudd — APPROVED** [review]

**Pull request merged**